fabric-tag-extensions-v0/src/main/java/net/fabricmc/fabric/impl/tag/extension/TagFactoryImpl.java

```diff
@@ -0,0 +1,119 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Stopwatch;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import net.minecraft.resource.ResourceManager;
+import net.minecraft.resource.ServerResourceManager;
+import net.minecraft.server.Main;
+import net.minecraft.tag.RequiredTagList;
+import net.minecraft.tag.RequiredTagListRegistry;
+import net.minecraft.tag.ServerTagManagerHolder;
+import net.minecraft.tag.Tag;
+import net.minecraft.tag.TagGroup;
+import net.minecraft.tag.TagGroupLoader;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.dynamic.RegistryOps;
+import net.minecraft.util.registry.DynamicRegistryManager;
+import net.minecraft.util.registry.Registry;
+import net.minecraft.util.registry.RegistryKey;
+
+import net.fabricmc.fabric.api.tag.TagFactory;
+import net.fabricmc.fabric.mixin.tag.extension.DynamicRegistryManagerAccessor;
+
+@SuppressWarnings("ClassCanBeRecord")
+public final class TagFactoryImpl<T> implements TagFactory<T> {
+	private static final Logger LOGGER = LogManager.getLogger();
+
+	public static final Map<RegistryKey<? extends Registry<?>>, RequiredTagList<?>> TAG_LISTS = new HashMap<>();
+	public static final Set<RequiredTagList<?>> DYNAMICS = new HashSet<>();
+
+	public static <T> TagFactory<T> of(Supplier<TagGroup<T>> tagGroupSupplier) {
+		return new TagFactoryImpl<>(tagGroupSupplier);
+	}
+
+	@SuppressWarnings("unchecked")
+	public static <T> TagFactory<T> of(RegistryKey<? extends Registry<T>> registryKey, String dataType) {
+		RequiredTagList<T> tagList;
+
+		// Use already registered tag list for the registry if it has the same dataType, in case multiple mods tried to do it.
+		if (TAG_LISTS.containsKey(registryKey)) {
+			tagList = (RequiredTagList<T>) TAG_LISTS.get(registryKey);
+			// Throw an exception if the tagList has different dataType.
+			Preconditions.checkArgument(tagList.getDataType().equals(dataType), "Tag list for registry %s is already existed with data type %s", registryKey.getValue(), tagList.getDataType());
+		} else {
+			tagList = RequiredTagListRegistry.register(registryKey, dataType);
+			TAG_LISTS.put(registryKey, tagList);
+
+			// Check whether the registry dynamic.
+			if (DynamicRegistryManagerAccessor.getInfos().containsKey(registryKey)) {
+				DYNAMICS.add(tagList);
+			}
+		}
+
+		return of(tagList::getGroup);
+	}
+
+	/**
+	 * Manually load tags for dynamic registries and add the resulting tag group to the tag list.
+	 *
+	 * <p>Minecraft loads the resource manager before dynamic registries, making tags for them failed to load
+	 * if it mentions datapack entries. The solution is to manually load tags after the registry is loaded.
+	 *
+	 * <p>Look at server's {@link Main#main} function calls for {@link ServerResourceManager#reload} and
+	 * {@link RegistryOps#method_36574} for the relevant code.
```

Comment:
Use `@see`

Reply:
IMO the current paragraph is more clear than using `@see`, `Main#main` is a huge function and we want it to be clear what to look for.
![image](https://user-images.githubusercontent.com/21150434/129130913-c8a273bc-c8d0-4aa1-bb1b-4dc47c3687a0.png)